File: src/api/servers.ts
import { call } from './client';
import { RustCommand } from './commands';
import type { Server, ServerInput } from './types/server.types';

export function getAllServers(): Promise<Server[]> {
  return call<Server[]>(RustCommand.GetAllServers);
}

export function getServer(id: number): Promise<Server> {
  return call<Server>(RustCommand.GetServer, { id });
}

export function createServer(input: ServerInput): Promise<Server> {
  return call<Server>(RustCommand.CreateServer, { input });
}

/** Backend drops the server's open connections on update */
export function updateServer(id: number, input: ServerInput): Promise<Server> {
  return call<Server>(RustCommand.UpdateServer, { id, input });
}

/** Deletes keychain password and open connections too */
export function deleteServer(id: number): Promise<void> {
  return call<void>(RustCommand.DeleteServer, { id });
}

File: src-tauri/src/commands/servers.rs
use tauri::State;

use crate::models::{Server, ServerInput};
use crate::state::AppState;
use crate::storage::repositories::servers;

#[tauri::command]
pub fn get_all_servers(state: State<'_, AppState>) -> Result<Vec<Server>, String> {
    servers::get_all(&state.storage).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn get_server(state: State<'_, AppState>, id: i64) -> Result<Server, String> {
    servers::get_by_id(&state.storage, id).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn create_server(state: State<'_, AppState>, input: ServerInput) -> Result<Server, String> {
    servers::create(&state.storage, input).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn update_server(
    state: State<'_, AppState>,
    id: i64,
    input: ServerInput,
) -> Result<Server, String> {
    // Disconnect existing connections when server config changes
    state.connections.disconnect_server(id);

    servers::update(&state.storage, id, input).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn delete_server(state: State<'_, AppState>, id: i64) -> Result<(), String> {
    // Disconnect all connections for this server first
    state.connections.disconnect_server(id);

    servers::delete(&state.storage, id).map_err(|e| e.to_string())
}